lib/elixir/lib/enum.ex

```diff
@@ -3821,18 +3821,23 @@ defmodule Enum do
 end
 
 defimpl Enumerable, for: List do
+  def count([]), do: {:ok, 0}
   def count(_list), do: {:error, __MODULE__}
+
+  def member?([], _value), do: {:ok, false}
   def member?(_list, _value), do: {:error, __MODULE__}
+
+  def slice([]), do: {:ok, 0, fn _, _ -> [] end}
   def slice(_list), do: {:error, __MODULE__}
 
   def reduce(_list, {:halt, acc}, _fun), do: {:halted, acc}
   def reduce(list, {:suspend, acc}, fun), do: {:suspended, acc, &reduce(list, &1, fun)}
   def reduce([], {:cont, acc}, _fun), do: {:done, acc}
   def reduce([head | tail], {:cont, acc}, fun), do: reduce(tail, fun.(head, acc), fun)
 
-  @doc false
   def slice(_list, _start, 0, _size), do: []
-  def slice(list, start, count, size) when start + count == size, do: list |> drop(start)
+  def slice(_list, _start, _count, 0), do: []
```

Comment:
I don't think this clause will ever be called, so you can remove it. I.e. it is not expected for callers to actually ask for a list of size zero.

Reply:
I think size is 0 when enumerable is empty, but it is caught by the previous clause (count == 0)

> Here are my results:
> 
> ```
> ##### With input list_empty #####
> Name                                     ips        average  deviation         median         99th %
> patched: Enum.Patched.empty?/1        7.35 M      136.02 ns   ±291.39%           0 ns        1000 ns
> original: Enum.empty?/1               7.29 M      137.14 ns   ±306.66%           0 ns        1000 ns
> 
> Comparison: 
> patched: Enum.Patched.empty?/1        7.35 M
> original: Enum.empty?/1               7.29 M - 1.01x slower +1.12 ns
> 
> ##### With input list_100 #####
> Name                                     ips        average  deviation         median         99th %
> patched: Enum.Patched.empty?/1       27.26 M       36.69 ns  ±1329.86%           0 ns         100 ns
> original: Enum.empty?/1               6.95 M      143.90 ns  ±4293.46%           0 ns        1000 ns
> 
> Comparison: 
> patched: Enum.Patched.empty?/1       27.26 M
> original: Enum.empty?/1               6.95 M - 3.92x slower +107.21 ns
> 
> ##### With input list_10_000 #####
> Name                                     ips        average  deviation         median         99th %
> original: Enum.empty?/1              42.49 M       23.53 ns   ±151.95%          20 ns          30 ns
> patched: Enum.Patched.empty?/1        7.30 M      136.91 ns   ±929.31%           0 ns        1000 ns
> 
> Comparison: 
> original: Enum.empty?/1              42.49 M
> patched: Enum.Patched.empty?/1        7.30 M - 5.82x slower +113.38 ns
> 
> ##### With input map_empty #####
> Name                                     ips        average  deviation         median         99th %
> original: Enum.empty?/1               4.43 M      225.96 ns  ±7620.11%           0 ns        1000 ns
> patched: Enum.Patched.empty?/1        4.04 M      247.28 ns  ±7222.88%           0 ns        1000 ns
> 
> Comparison: 
> original: Enum.empty?/1               4.43 M
> patched: Enum.Patched.empty?/1        4.04 M - 1.09x slower +21.32 ns
> 
> ##### With input map_100 #####
> Name                                     ips        average  deviation         median         99th %
> original: Enum.empty?/1               4.11 M      243.33 ns  ±9368.85%           0 ns        1000 ns
> patched: Enum.Patched.empty?/1        4.00 M      249.74 ns  ±9615.02%           0 ns        1000 ns
> 
> Comparison: 
> original: Enum.empty?/1               4.11 M
> patched: Enum.Patched.empty?/1        4.00 M - 1.03x slower +6.41 ns
> 
> ##### With input map_10_000 #####
> Name                                     ips        average  deviation         median         99th %
> patched: Enum.Patched.empty?/1        4.24 M      235.95 ns  ±5571.78%           0 ns        1000 ns
> original: Enum.empty?/1               4.24 M      236.01 ns  ±5642.25%           0 ns        1000 ns
> 
> Comparison: 
> patched: Enum.Patched.empty?/1        4.24 M
> original: Enum.empty?/1               4.24 M - 1.00x slower +0.0566 ns
> ```
> 
> I think the code is already hitting all fast branches today so that it is so fast that measurements cannot be reliable. As mentioned above, looking at the code, I can't see when the slice function is invoked when the size is zero. As benchee puts it:
> 
> > Warning: The function you are trying to benchmark is super fast, making measurements more unreliable!
> 
> In any case, thanks for the PR!



> Here are my results:
> 
> ```
> ##### With input list_empty #####
> Name                                     ips        average  deviation         median         99th %
> patched: Enum.Patched.empty?/1        7.35 M      136.02 ns   ±291.39%           0 ns        1000 ns
> original: Enum.empty?/1               7.29 M      137.14 ns   ±306.66%           0 ns        1000 ns
> 
> Comparison: 
> patched: Enum.Patched.empty?/1        7.35 M
> original: Enum.empty?/1               7.29 M - 1.01x slower +1.12 ns
> 
> ##### With input list_100 #####
> Name                                     ips        average  deviation         median         99th %
> patched: Enum.Patched.empty?/1       27.26 M       36.69 ns  ±1329.86%           0 ns         100 ns
> original: Enum.empty?/1               6.95 M      143.90 ns  ±4293.46%           0 ns        1000 ns
> 
> Comparison: 
> patched: Enum.Patched.empty?/1       27.26 M
> original: Enum.empty?/1               6.95 M - 3.92x slower +107.21 ns
> 
> ##### With input list_10_000 #####
> Name                                     ips        average  deviation         median         99th %
> original: Enum.empty?/1              42.49 M       23.53 ns   ±151.95%          20 ns          30 ns
> patched: Enum.Patched.empty?/1        7.30 M      136.91 ns   ±929.31%           0 ns        1000 ns
> 
> Comparison: 
> original: Enum.empty?/1              42.49 M
> patched: Enum.Patched.empty?/1        7.30 M - 5.82x slower +113.38 ns
> 
> ##### With input map_empty #####
> Name                                     ips        average  deviation         median         99th %
> original: Enum.empty?/1               4.43 M      225.96 ns  ±7620.11%           0 ns        1000 ns
> patched: Enum.Patched.empty?/1        4.04 M      247.28 ns  ±7222.88%           0 ns        1000 ns
> 
> Comparison: 
> original: Enum.empty?/1               4.43 M
> patched: Enum.Patched.empty?/1        4.04 M - 1.09x slower +21.32 ns
> 
> ##### With input map_100 #####
> Name                                     ips        average  deviation         median         99th %
> original: Enum.empty?/1               4.11 M      243.33 ns  ±9368.85%           0 ns        1000 ns
> patched: Enum.Patched.empty?/1        4.00 M      249.74 ns  ±9615.02%           0 ns        1000 ns
> 
> Comparison: 
> original: Enum.empty?/1               4.11 M
> patched: Enum.Patched.empty?/1        4.00 M - 1.03x slower +6.41 ns
> 
> ##### With input map_10_000 #####
> Name                                     ips        average  deviation         median         99th %
> patched: Enum.Patched.empty?/1        4.24 M      235.95 ns  ±5571.78%           0 ns        1000 ns
> original: Enum.empty?/1               4.24 M      236.01 ns  ±5642.25%           0 ns        1000 ns
> 
> Comparison: 
> patched: Enum.Patched.empty?/1        4.24 M
> original: Enum.empty?/1               4.24 M - 1.00x slower +0.0566 ns
> ```
> 
> I think the code is already hitting all fast branches today so that it is so fast that measurements cannot be reliable. As mentioned above, looking at the code, I can't see when the slice function is invoked when the size is zero. As benchee puts it:
> 
> > Warning: The function you are trying to benchmark is super fast, making measurements more unreliable!
> 
> In any case, thanks for the PR!

The results did vary a lot, but on my end  the Patched version always outperformed the original one, I never got the Banchee warning though, probably do tue my computer not being that fast.